libs/@local/harpc/tower/src/body/mod.rs

```diff
@@ -0,0 +1,309 @@
+pub mod boxed;
+pub mod controlled;
+pub mod empty;
+pub mod full;
+pub mod limited;
+pub mod map;
+pub mod server;
+pub mod size_hint;
+pub mod stream;
+pub mod timeout;
+
+use core::{
+    ops::DerefMut,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::Buf;
+
+pub use self::size_hint::SizeHint;
+use self::{
+    boxed::{BoxBody, UnsyncBoxBody},
+    map::{MapControl, MapData, MapError},
+    stream::BodyStream,
+};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Frame<D, C> {
+    Data(D),
+    Control(C),
+}
+
+impl<D, C> Frame<D, C> {
+    pub const fn new_data(data: D) -> Self {
+        Self::Data(data)
+    }
+
+    pub const fn new_control(control: C) -> Self {
+        Self::Control(control)
+    }
+
+    pub const fn data(&self) -> Option<&D> {
+        match self {
+            Self::Data(data) => Some(data),
+            Self::Control(_) => None,
+        }
+    }
+
+    pub fn data_mut(&mut self) -> Option<&mut D> {
+        match self {
+            Self::Data(data) => Some(data),
+            Self::Control(_) => None,
+        }
+    }
+
+    pub fn map_data<D2>(self, func: impl FnOnce(D) -> D2) -> Frame<D2, C> {
+        match self {
+            Self::Data(data) => Frame::new_data(func(data)),
+            Self::Control(control) => Frame::new_control(control),
+        }
+    }
+
+    pub fn into_data(self) -> Result<D, Self> {
+        match self {
+            Self::Data(data) => Ok(data),
+            Self::Control(control) => Err(Self::new_control(control)),
+        }
+    }
+
+    pub const fn control(&self) -> Option<&C> {
+        match self {
+            Self::Data(_) => None,
+            Self::Control(control) => Some(control),
+        }
+    }
+
+    pub fn control_mut(&mut self) -> Option<&mut C> {
+        match self {
+            Self::Data(_) => None,
+            Self::Control(control) => Some(control),
+        }
+    }
+
+    pub fn map_control<C2>(self, func: impl FnOnce(C) -> C2) -> Frame<D, C2> {
+        match self {
+            Self::Data(data) => Frame::new_data(data),
+            Self::Control(control) => Frame::new_control(func(control)),
+        }
+    }
+
+    pub fn into_control(self) -> Result<C, Self> {
+        match self {
+            Self::Data(data) => Err(Self::new_data(data)),
+            Self::Control(control) => Ok(control),
+        }
+    }
+}
+
+pub(crate) type BodyFrame<B> = Frame<<B as Body>::Data, <B as Body>::Control>;
+pub(crate) type BodyFrameResult<B> = Result<BodyFrame<B>, <B as Body>::Error>;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum BodyState {
+    Complete,
+    Incomplete,
+}
+
+pub trait Body {
+    type Control;
+    type Data: Buf;
+
+    type Error;
+
+    fn poll_frame(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<BodyFrameResult<Self>>>;
+
+    /// Signal if the body has been is complete.
```

Comment:
```suggestion
    /// Signal if the body is complete.
```

Reply:
What is a "complete body"?

Reply:
elaborated on in [28d6155](https://github.com/hashintel/hash/pull/4713/commits/28d6155d99f98bf93d7c75a8f44e880900a488c0)